server/app/controllers/admin/AdminProgramStatusesController.java

```diff
@@ -34,6 +57,210 @@ public Result index(Http.Request request, long programId) throws ProgramNotFound
+      return notFound("status tracking is not enabled");
+    }
+    requestChecker.throwIfProgramNotDraft(programId);
+    ProgramDefinition program = service.getProgramDefinition(programId);
+
+    DynamicForm requestData = formFactory.form().bindFromRequest(request);
+    String rawStatusText = requestData.get(ProgramStatusesView.DELETE_STATUS_TEXT_NAME);
+    if (Strings.isNullOrEmpty(rawStatusText)) {
+      return badRequest("missing or empty status text");
+    }
+    final String statusText = rawStatusText.trim();
+
+    StatusDefinitions current = program.statusDefinitions();
+    try {
+      current.setStatuses(removeStatus(current.getStatuses(), statusText));
+    } catch (MissingStatusException e) {
+      return redirect(routes.AdminProgramStatusesController.index(programId).url())
+          .flashing(
+              "error",
+              "The status being deleted no longer exists and may have been deleted in a"
+                  + " separate window.");
+    }
+    ErrorAnd<ProgramDefinition, CiviFormError> setStatusResult =
+        service.setStatuses(programId, current);
+    Result result =
+        redirect(routes.AdminProgramStatusesController.index(programId).url())
+            .flashing("success", "Status deleted");
+    if (setStatusResult.isError()) {
+      String errorMessage = joinErrors(setStatusResult.getErrors());
+      result = result.flashing("error", errorMessage);
+    } else {
+      result = result.flashing("success", "Status deleted");
+    }
+
+    return result;
+  }
+
+  private static ImmutableList<StatusDefinitions.Status> addStatus(
+      ImmutableList<StatusDefinitions.Status> statuses, StatusDefinitions.Status newStatus)
+      throws DuplicateStatusException {
+    if (statuses.stream()
+        .filter(s -> s.statusText().equals(newStatus.statusText()))
+        .findAny()
+        .isPresent()) {
+      throw new DuplicateStatusException(newStatus.statusText());
+    }
+    return ImmutableList.<StatusDefinitions.Status>builder()
+        .addAll(statuses)
+        .add(newStatus)
+        .build();
+  }
+
+  private static ImmutableList<StatusDefinitions.Status> replaceStatus(
+      ImmutableList<StatusDefinitions.Status> statuses,
+      String replaceStatusName,
+      Function<StatusDefinitions.Status, StatusDefinitions.Status> replacerFunc)
+      throws MissingStatusException, DuplicateStatusException {
+    ImmutableMap<String, Integer> statusNameToIndex = statusNameToIndexMap(statuses);
+    if (!statusNameToIndex.containsKey(replaceStatusName)) {
+      throw new MissingStatusException();
+    }
+    StatusDefinitions.Status editedStatus =
+        replacerFunc.apply(statuses.get(statusNameToIndex.get(replaceStatusName)));
+    if (!replaceStatusName.equals(editedStatus.statusText())
+        && statusNameToIndex.containsKey(editedStatus.statusText())) {
+      throw new DuplicateStatusException(editedStatus.statusText());
+    }
+    return IntStream.range(0, statuses.size())
```

Comment:
I think it's a lot cleaner to just copy the ImmutableList to a List, replace the index, then convert back